### ISSUE TITLE
Synchronizing Playing Multiple Audio

### ISSUE THREAD

**Issue by shoumo680** [google/ExoPlayer#4261]
Hello, @ojw28 

### Issue description
I want to use ExoPlayer broadcast six audio, I use the #2855, the difference is I use HlsMediaSource, although it can run, but has a chance to play out of sync, more obvious after seekTo. Ts fragment is 10 seconds at the beginning, change to 4 seconds, less incidence of the problem. I have reference #4014  #4051, but couldn't find a solution.

Thanks in advance!

### Version of ExoPlayer being used
2.7.3

**Comment by akwolf** [comment]
i have the same problem ,if i donn't drag the seekBar forward the progress the player work perfect,when i drag the seekBar forward then  audio's playing order gets confusing.i struggled for serial days and attempt many ways, Don't know what the problem is?Eagerly awaiting your reply,thks!
Here's what I did.

**1、i create a RenderersFactory implements DefaultRenderersFactory and my own MediaCodecAudioRenderer**
```java
public class MultiTrackRenderersFactory extends DefaultRenderersFactory {
    private int audioTrackCnt;

    public MultiTrackRenderersFactory(int audioTrackCnt, Context context, @Nullable DrmSessionManager<FrameworkMediaCrypto> drmSessionManager, int extensionRendererMode) {
        super(context, drmSessionManager, extensionRendererMode);
        this.audioTrackCnt = audioTrackCnt;
    }

    @Override
    protected void buildAudioRenderers(Context context, @Nullable DrmSessionManager<FrameworkMediaCrypto> drmSessionManager, AudioProcessor[] audioProcessors, Handler eventHandler, AudioRendererEventListener eventListener, int extensionRendererMode, ArrayList<Renderer> out) {
        for (int i = 0; i < audioTrackCnt; i++) {
            out.add(new MultiTrackCodecAudioRenderer(i, MediaCodecSelector.DEFAULT, drmSessionManager, true, eventHandler, eventListener, AudioCapabilities.getCapabilities(context), audioProcessors));
        }
    }
}
```
 
```java
public class MultiTrackCodecAudioRenderer extends MediaCodecAudioRenderer {

    private int index;

    public MultiTrackCodecAudioRenderer(int index, MediaCodecSelector mediaCodecSelector, @Nullable DrmSessionManager<FrameworkMediaCrypto> drmSessionManager, boolean playClearSamplesWithoutKeys, @Nullable Handler eventHandler, @Nullable AudioRendererEventListener eventListener, @Nullable AudioCapabilities audioCapabilities, AudioProcessor... audioProcessors) {
        super(mediaCodecSelector, drmSessionManager, playClearSamplesWithoutKeys, eventHandler, eventListener, audioCapabilities, audioProcessors);
        this.index = index;
    }
    @Override
    public MediaClock getMediaClock() {
        Log.i(Const.TAG, "getMediaClock:" + String.valueOf(super.getMediaClock().getPositionUs()));
        if (index == 0) {
            return super.getMediaClock();
        }
        return null;
    }
}
```

**2、in the main Activity  combine all HlsMediaSource in one MergingMediaSource**
```java
private void initHLSPlayer() {
        MediaSource sampleSource1 = new HlsMediaSource(Uri.parse("http://ybproduct.yourband.net/820959723363983360.m3u8"), dataSourceFactory, handler, this);
        MediaSource sampleSource2 = new HlsMediaSource(Uri.parse("http://ybproduct.yourband.net/820959445826887680.m3u8"), dataSourceFactory, handler, this);
        MediaSource sampleSource3 = new HlsMediaSource(Uri.parse("http://ybproduct.yourband.net/820959551737257984.m3u8"), dataSourceFactory, handler, this);
        MediaSource sampleSource4 = new HlsMediaSource(Uri.parse("http://ybproduct.yourband.net/820959497194528768.m3u8"), dataSourceFactory, handler, this);
        MediaSource sampleSource5 = new HlsMediaSource(Uri.parse("http://ybproduct.yourband.net/820959643940642816.m3u8"), dataSourceFactory, handler, this);
        MergingMediaSource mediaSource = new MergingMediaSource(sampleSource1, sampleSource2, sampleSource3, sampleSource4,sampleSource5);
        initExoPlayer(mediaSource);
    }
```

**3、at the final write MultiAudioTrackSelector implements TrackSelector for assign TrackSelection for each track**

```java
public class MultiAudioTrackSelector extends TrackSelector {
    
    @Override
    public TrackSelectorResult selectTracks(RendererCapabilities[] rendererCapabilities, TrackGroupArray trackGroups) throws ExoPlaybackException {

        int rendererCount = rendererCapabilities.length;
        TrackSelection[] rendererTrackSelections = new TrackSelection[rendererCount];
        int idx = 0;
        for (int i = 0; i < rendererCount; i++) {
            int trackType = rendererCapabilities[i].getTrackType();
            int length = trackGroups.length;
            if (trackType == C.TRACK_TYPE_AUDIO) {
                int trackGroup = 2 * idx + 1;
                rendererTrackSelections[i] = new FixedTrackSelection(trackGroups.get(trackGroup), 0);
                idx++;
            }
        }

        RendererConfiguration[] rendererConfigurations = new RendererConfiguration[rendererCapabilities.length];
        for (int i = 0; i < rendererCount; i++) {
            rendererConfigurations[i] = rendererTrackSelections[i] != null ? RendererConfiguration.DEFAULT : null;
        }
        return new TrackSelectorResult(trackGroups, new TrackSelectionArray(rendererTrackSelections), null, rendererConfigurations);
    }

    @Override
    public void onSelectionActivated(Object info) {

    }
}
```

**Comment by AquilesCanta** [comment]
Hello, there are two issues to solve for synchronizing multiple audio tracks:
- The first is ensuring that HlsMediaSources are outputting the samples so that synchronization works as intended. This depends a lot on what media looks like. If it is VOD, it might just work. But I suppose it makes sense to solve the next item first.
- The second is actually rendering them in sync. @andrewlewis suggests it might be possible to do using just AudioTrack, but that the only way to achieve perfect synchronization is probably using a mix of custom AudioProcessors / AudioSink.

@akwolf, in your case you are just not using any means of synchronization, as far as I can tell, considering there is no shared clock between renderers. Basically there is a race between all renderers (getMediaClock returns null for all except the first). What I think could be done:

- If needed, normalize the sample rate.
------ From here onwards, things should probably be done by a shared audio sink -----
- Wait until you have samples to render for everyone of the renderers.
- Mix everything into a single buffer.
- Output this single buffer (could be forwarded to DefaultAudioSink).

@andrewlewis, please correct anything that looks wrong (would using only AudioTrack be simpler?). @shoumo680, I am going to assume you are doing something similar to what @akwolf is doing. That is, using a merging media source (I don't know about the rendering part).

**Comment by shoumo680** [comment]
Yes, I'm doing the same thing to @akwolf .

I updated it to 2.8.0 version .

 **From here onwards, things should probably be done by a shared audio sink** ----Is it all Renderer shared an AudioSink?but DefaultAudioSink method handleBuffer line 615 throw  exception.

**Wait until you have samples to render for everyone of the renderers.
Mix everything into a single buffer.
Output this single buffer (could be forwarded to DefaultAudioSink).**----Sorry, I don't quite understand how to do this?

**Comment by andrewlewis** [comment]
The solution to rendering that @AquilesCanta outlined sounds right:
- You should have one "master" audio renderer provide the media clock, and this should have a `DefaultAudioSink` for playing audio. This master renderer also needs to care of mixing together its own audio stream with the streams from all the other renderers.
- One way to do that is to pass it a custom mixing `AudioProcessor` which you'll need to implement, that takes care of getting samples from all the different audio renderers and mixing them together, outputting a buffer only once all corresponding input buffers have been received.
- The other audio renderers can have custom `AudioSink`s which take care of receiving audio buffers and passing those to your mixing `AudioProcessor`.

As you can probably tell from the description, this is going to be quite a lot of work to implement, and it's not really feasible for us to provide detailed guidance. I think any solution with perfect (sample-accurate) synchronization of your audio streams needs to be using just one underlying platform `AudioTrack`.

If you can compromise on perfect synchronization, you may be able to achieve a similar effect using several `MediaCodecAudioRenderer`s, each with their own output track. These may drift out of synchronization a bit. However, it sounds like you tried to do this and it didn't work. I'm not sure there's enough information here to diagnose what's going on, though. I suggest adding some logging to see what the current position returned by each audio renderer is after seeking, so you can investigate why there is a discrepancy.

**Comment by shoumo680** [comment]
I will try your proposal.
   
  It seems that the getPositionUs of the Renderer other than the main Renderer is not called.
  This is the log record that I actively call getPositionUs .
   
   **seekTo before**
   `getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---1319488
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---1300075
    getPositionUs: MediaCodecAudioRenderer@78972a6---getPositionUs---1300159
    getPositionUs: MediaCodecAudioRenderer@8d894e7---getPositionUs---1300235
    getPositionUs: MediaCodecAudioRenderer@33ff894---getPositionUs---1260361
    getPositionUs: MediaCodecAudioRenderer@6721f3d---getPositionUs---1280458
    getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---1320102
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---1308707
    getPositionUs: MediaCodecAudioRenderer@78972a6---getPositionUs---1308883
    getPositionUs: MediaCodecAudioRenderer@8d894e7---getPositionUs---1308964
    getPositionUs: MediaCodecAudioRenderer@33ff894---getPositionUs---1269119
    getPositionUs: MediaCodecAudioRenderer@6721f3d---getPositionUs---1289214
    getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---1328718
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---1309304
    getPositionUs: MediaCodecAudioRenderer@78972a6---getPositionUs---1309375
    getPositionUs: MediaCodecAudioRenderer@8d894e7---getPositionUs---1309444`

    **seekTo later**
 
 `getPositionUs: MediaCodecAudioRenderer@33ff894---getPositionUs---127815827
    getPositionUs: MediaCodecAudioRenderer@6721f3d---getPositionUs---127796014
    getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---127815105
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---127996397
    getPositionUs: MediaCodecAudioRenderer@78972a6---getPositionUs---127816630
    getPositionUs: MediaCodecAudioRenderer@8d894e7---getPositionUs---127816840
    getPositionUs: MediaCodecAudioRenderer@33ff894---getPositionUs---127817009
    getPositionUs: MediaCodecAudioRenderer@6721f3d---getPositionUs---127797253
    getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---127816279
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---128005476
    getPositionUs: MediaCodecAudioRenderer@78972a6---getPositionUs---127825774
    getPositionUs: MediaCodecAudioRenderer@8d894e7---getPositionUs---127825930
    getPositionUs: MediaCodecAudioRenderer@33ff894---getPositionUs---127826055
    getPositionUs: MediaCodecAudioRenderer@6721f3d---getPositionUs---127806208
    getPositionUs: MediaCodecAudioRenderer@a4a1332---getPositionUs---127825417
    getPositionUs: MediaCodecAudioRenderer@fe64401---getPositionUs---128006474
`

**Comment by andrewlewis** [comment]
The player will only poll `getPositionUs` on the `Renderer` that's providing the media clock. It looks like the positions are close. If you're looking for finer synchronization than a few hundred milliseconds then I think you need to take the mixing approach described above.

Could you give more detail on what is actually going wrong? Do you hear audio from completely different parts of the stream from the different renderers after seeking? If you play the same URL with every source are they synchronized?

**Comment by shoumo680** [comment]
Play the audio of six different instrument, seekTo after one or two audio probability not synchronous,again seekTo will synchronized. the Ts the length of the reduced to 4 seconds will reduce the probability.The use of Hls will appear this problem, and with ExtractorMediaSource Mp3 does not appear.

**Comment by andrewlewis** [comment]
@AquilesCanta It sounds like the problem is worse with longer HLS chunks. Any ideas what could be going on here?

**Comment by shoumo680** [comment]
The encrypted HLS and the encrypted mp3 will be out of sync. Is this problem related to encryption?

**Comment by AquilesCanta** [comment]
> Is this problem related to encryption

Not likely, no.

**Comment by darkkillen** [comment]
I have a similar problem. I try to mixing the audio prior to feeding it into a single AudioTrack instance. as #2200 but It's has a lot of work. Can you suggest any idea to achieve this?

**Comment by andrewlewis** [comment]
This thread describes a couple of possible ways to synchronize playing multiple separate audio streams but they are quite a lot of work. As this doesn't seem to be a popular use case I'm afraid we're not planning to support it, either by adding functionality to the player or by giving more guidance on how to implement it.

It sounds like there's also an issue that might be related to HLS but not enough information here to diagnose it. Please file a new issue if follow-up is still required on that. Thanks.
